src/array_borsh.rs

```diff
@@ -0,0 +1,101 @@
+use crate::imp_prelude::*;
+use crate::IntoDimension;
+use alloc::vec::Vec;
+use borsh::{BorshDeserialize, BorshSerialize};
+use core::ops::Deref;
+
+/// **Requires crate feature `"borsh"`**
+impl<I> BorshSerialize for Dim<I>
+where
+    I: BorshSerialize,
+{
+    fn serialize<W: borsh::io::Write>(&self, writer: &mut W) -> borsh::io::Result<()> {
+        <I as BorshSerialize>::serialize(&self.ix(), writer)
+    }
+}
+
+/// **Requires crate feature `"borsh"`**
+impl<I> BorshDeserialize for Dim<I>
+where
+    I: BorshDeserialize,
+{
+    fn deserialize_reader<R: borsh::io::Read>(reader: &mut R) -> borsh::io::Result<Self> {
+        <I as BorshDeserialize>::deserialize_reader(reader).map(Dim::new)
+    }
+}
+
+/// **Requires crate feature `"borsh"`**
+impl BorshSerialize for IxDyn {
+    fn serialize<W: borsh::io::Write>(&self, writer: &mut W) -> borsh::io::Result<()> {
+        let elts = self.ix().deref();
+        // Output length of dimensions.
+        <usize as BorshSerialize>::serialize(&elts.len(), writer)?;
+        // Followed by actual data.
+        for elt in elts {
+            <Ix as BorshSerialize>::serialize(elt, writer)?;
+        }
+        Ok(())
+    }
+}
+
+/// **Requires crate feature `"borsh"`**
+impl BorshDeserialize for IxDyn {
+    fn deserialize_reader<R: borsh::io::Read>(reader: &mut R) -> borsh::io::Result<Self> {
+        // Deserialize the length.
+        let len = <usize as BorshDeserialize>::deserialize_reader(reader)?;
+        // Deserialize the given number of elements. We assume the source is
+        // trusted so we use a capacity hint...
+        let mut elts = Vec::with_capacity(len);
+        for _ix in 0..len {
+            elts.push(<Ix as BorshDeserialize>::deserialize_reader(reader)?);
+        }
+        Ok(elts.into_dimension())
+    }
+}
+
+/// **Requires crate feature `"borsh"`**
+impl<A, D, S> BorshSerialize for ArrayBase<S, D>
+where
+    A: BorshSerialize,
+    D: Dimension + BorshSerialize,
+    S: Data<Elem = A>,
+{
+    fn serialize<W: borsh::io::Write>(&self, writer: &mut W) -> borsh::io::Result<()> {
+        // Dimensions
+        <D as BorshSerialize>::serialize(&self.raw_dim(), writer)?;
+        // Followed by length of data
+        let iter = self.iter();
+        <usize as BorshSerialize>::serialize(&iter.len(), writer)?;
+        // Followed by data itself.
+        for elt in iter {
+            <A as BorshSerialize>::serialize(elt, writer)?;
+        }
+        Ok(())
+    }
+}
+
+/// **Requires crate feature `"borsh"`**
+impl<A, D, S> BorshDeserialize for ArrayBase<S, D>
+where
+    A: BorshDeserialize,
+    D: BorshDeserialize + Dimension,
+    S: DataOwned<Elem = A>,
+{
+    fn deserialize_reader<R: borsh::io::Read>(reader: &mut R) -> borsh::io::Result<Self> {
+        // Dimensions
+        let dim = <D as BorshDeserialize>::deserialize_reader(reader)?;
+        // Followed by length of data
+        let len = <usize as BorshDeserialize>::deserialize_reader(reader)?;
```

Comment:
How do you handle integer size here? What if it was serialized on a 64-bit usize platform but you deserialize on a 32-bit?

Reply:
I would look at https://github.com/indexmap-rs/indexmap/pull/313 for inspiration - fixed size integers and check error cases

Reply:
borsh always serialises usize as u64: https://docs.rs/borsh/latest/src/borsh/ser/mod.rs.html#128-132

Reply:
borsh says that container sizes should be stored as u32. Unsure how it interacts with ndarray. For me, developing this PR is not my priority unfortunately. Help from others welcome.